cranelift-codegen/meta/src/cdsl/instructions.rs

```diff
@@ -406,36 +418,102 @@ impl ValueTypeOrAny {
     }
 }
 
+type VectorBitWidth = u64;
+pub enum InstructionParameter {
+    AnyType,
+    LaneType(LaneType),
+    VectorType(LaneType, VectorBitWidth),
+    ReferenceType(ReferenceType),
+}
+
+impl From<shared::types::Int> for InstructionParameter {
+    fn from(ty: shared::types::Int) -> Self {
+        InstructionParameter::LaneType(ty.into())
+    }
+}
+
+impl From<shared::types::Bool> for InstructionParameter {
+    fn from(ty: shared::types::Bool) -> Self {
+        InstructionParameter::LaneType(ty.into())
+    }
+}
+
+impl From<shared::types::Float> for InstructionParameter {
+    fn from(ty: shared::types::Float) -> Self {
+        InstructionParameter::LaneType(ty.into())
+    }
+}
+
+impl From<cdsl::types::LaneType> for InstructionParameter {
+    fn from(ty: cdsl::types::LaneType) -> Self {
+        InstructionParameter::LaneType(ty.into())
+    }
+}
+
 #[derive(Clone)]
 pub struct BoundInstruction {
     pub inst: Instruction,
     pub value_types: Vec<ValueTypeOrAny>,
 }
 
 impl BoundInstruction {
-    pub fn bind(self, lane_type: impl Into<LaneType>) -> BoundInstruction {
-        bind(self.inst, Some(lane_type.into()), self.value_types)
-    }
-
-    pub fn bind_ref(self, reference_type: impl Into<ReferenceType>) -> BoundInstruction {
-        bind_ref(self.inst, Some(reference_type.into()), self.value_types)
+    /// Construct a new bound instruction (with nothing bound yet) from an instruction
+    fn new(inst: &Instruction) -> Self {
+        BoundInstruction {
+            inst: inst.clone(),
+            value_types: vec![],
+        }
     }
 
-    pub fn bind_vector_from_lane(
-        self,
-        lane_type: impl Into<LaneType>,
-        vector_size_in_bits: u64,
-    ) -> BoundInstruction {
-        bind_vector(
-            self.inst,
-            lane_type.into(),
-            vector_size_in_bits,
-            self.value_types,
-        )
+    /// Verify that binding types to the instruction does not violate the polymorphic rules.
+    fn verify_polymorphic_binding(&self) -> Result<(), String> {
+        match &self.inst.polymorphic_info {
+            Some(poly) => {
+                if self.value_types.len() <= 1 + poly.other_typevars.len() {
+                    Ok(())
+                } else {
+                    Err(format!(
+                        "trying to bind too many types for {}",
+                        self.inst.name
+                    ))
+                }
+            }
+            None => Err(format!(
+                "trying to bind a type for {} which is not a polymorphic instruction",
+                self.inst.name
+            )),
+        }
     }
+}
 
-    pub fn bind_any(self) -> BoundInstruction {
-        bind(self.inst, None, self.value_types)
+impl Bindable for BoundInstruction {
+    fn bind(&self, parameter: impl Into<InstructionParameter>) -> BoundInstruction {
+        let mut modified = self.clone();
+        match parameter.into() {
```

Comment:
Well, I can't currently because the last case (Immediate) pushes to a different array. I wonder if there's a way to keep all the types and immediates in the same array... this would be good for checking the order (I think) and then I could simplify this match.